internal/run/run.go

```diff
@@ -0,0 +1,310 @@
+package run
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"slices"
+	"strings"
+
+	"github.com/manifoldco/promptui"
+	"github.com/speakeasy-api/openapi-generation/v2/pkg/generate"
+	"github.com/speakeasy-api/sdk-gen-config/workflow"
+	"github.com/speakeasy-api/speakeasy/internal/config"
+	"github.com/speakeasy-api/speakeasy/internal/download"
+	"github.com/speakeasy-api/speakeasy/internal/overlay"
+	"github.com/speakeasy-api/speakeasy/internal/sdkgen"
+	"github.com/speakeasy-api/speakeasy/internal/utils"
+	"github.com/speakeasy-api/speakeasy/internal/validation"
+	"github.com/speakeasy-api/speakeasy/pkg/merge"
+)
+
+func Run(ctx context.Context, target, source, genVersion, installationURL, repo, repoSubDir string, debug bool) error {
+	workingDir, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	wf, _, err := workflow.Load(workingDir)
+	if err != nil {
+		return err
+	}
+
+	if err := wf.Validate(generate.GetSupportedLanguages()); err != nil {
+		return err
+	}
+
+	targets := []string{}
+	for targetID := range wf.Targets {
+		targets = append(targets, targetID)
+	}
+	slices.Sort(targets)
+
+	sources := []string{}
+	for sourceID := range wf.Sources {
+		sources = append(sources, sourceID)
+	}
+	slices.Sort(sources)
+
+	if target == "" && source == "" {
+		if len(wf.Targets) == 1 {
+			target = targets[0]
+		} else if len(wf.Targets) == 0 && len(wf.Sources) == 1 {
+			source = sources[0]
+		} else {
+			// TODO update to use our proper interactive code
+			prompt := promptui.Prompt{
+				Label: fmt.Sprintf("Select a target (%s or 'all')", strings.Join(targets, ", ")),
+				Validate: func(input string) error {
+					if input == "" {
+						return fmt.Errorf("target cannot be empty")
+					}
+
+					if input != "all" && !slices.Contains(targets, input) {
+						return fmt.Errorf("invalid target")
+					}
+
+					return nil
+				},
+			}
+
+			result, err := prompt.Run()
+			if err != nil {
+				return err
+			}
+
+			target = result
+		}
+	}
+
+	// TODO we maybe want to be smarter about doubling up work if someone specifies both a target and a source via flags
+	if source == "all" {
+		for id, s := range wf.Sources {
+			if _, err := runSource(ctx, id, &s); err != nil {
+				return err
+			}
+		}
+	}
+
+	if target == "all" {
+		for t := range wf.Targets {
+			if err := runTarget(ctx, t, wf, workingDir, genVersion, installationURL, repo, repoSubDir, debug); err != nil {
+				return err
+			}
+		}
+	}
+
+	if source != "" && source != "all" {
```

Comment:
There's a bit of weirdness here I think. I added two sources, `s1` and `s2`. I set up a target that referenced `s1`. If I run just `speakeasy run`, it runs `s1` and the target. If i run `speakeasy run -s s2` it just runs s2. If i run `speakeasy run -s s2 -t all` it runs just s1 and the target.

Imo the solution to this might be to not allow -s and -t to be specificified at the same time. If you are generating targets, we will determine what sources need to be generated. `-s` should be a special case to tell us to _only_ generate sources

Reply:
okay have dealt with this you can either specify a source or a target